sql: add validation for partitions matching REGIONAL BY ROW

otan opened the pull request:
This commit adds validation that partitions for a REGIONAL BY ROW table
matches regions in the database descriptor.

This required some hacking - the type schema changer modifies the enum
and the table separately. If we modify the enum before the table, the
validation would fail on fetching the table descriptor as the partitions
are not there. If we modify the table before the enum or in the same
batch, the new region is still in the adding state and would hit an
assertion error. To remedy this, we have to fetch the tables before the
enum is changed, then re-hydrate the type in the columns so that the new
region value is not read only.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63459)
<!-- Reviewable:end -->

otan commented:
```

E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +relation â€¹"rbr"â€º (55): invalid locality config: missing partition â€¹us-east2â€º on PRIMARY INDEX of table â€¹rbrâ€º
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +(1) keys: [sql.schema.validation_errors.read.cross_references.relation]
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +Wraps: (2) keys: [sql.schema.validation_errors.read.cross_references.relation]
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +Wraps: (3)
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | (opaque error wrapper)
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | reportable 0:
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  |
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.(*validationErrorAccumulator).decorate
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:299
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.(*validationErrorAccumulator).Report
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:244
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc.(*wrapper).ValidateCrossReferences
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc/validate.go:145
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.Validate.func2
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:65
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.(*validationErrorAccumulator).validateDescriptorsAtLevel
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:255
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.Validate
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:60
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv.descriptorFromKeyValue
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv/catalogkv.go:168
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv.getDescriptorByID
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv/catalogkv.go:420
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv.MustGetDescriptorByID
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv/catalogkv.go:510
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).WaitForOneVersion.func1
E210411 23:51:00.664965 35357 sql/sqltelemetry/report.go:57 â‹® [n1,client=â€¹127.0.0.1:54294â€º,hostssl,user=root] 6216 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:330
```

looks like we should block concurrency of ADD/DROP region.

otan committed:
sql: add validation for partitions matching REGIONAL BY ROW

This commit adds validation that partitions for a REGIONAL BY ROW table
matches regions in the database descriptor.

This required some hacking - the type schema changer modifies the enum
and the table separately. If we modify the enum before the table, the
validation would fail on fetching the table descriptor as the partitions
are not there. If we modify the table before the enum or in the same
batch, the new region is still in the adding state and would hit an
assertion error. To remedy this, we have to fetch the tables before the
enum is changed, then re-hydrate the type in the columns so that the new
region value is not read only.

Release note: None

arulajmani reviewed (commented):
After staring at `TestConcurrentAddDropRegions` for a bit I found that it is only failing if the first operation (the one that blocks) is a drop (concurrent-drops, concurrent drop-add). This is because the second operation re-generates the RegionConfig based on the set of PUBLIC regions, which is missing the region that was dropped by the first operation as it moved it to READ ONLY. A key aside here is that we [do not validate cross references when writing the newly partitioned descriptor here](https://github.com/cockroachdb/cockroach/blob/b33933e3c02eec9d49f5a401402a7a2916124e3e/pkg/sql/catalog/descs/collection.go#L1453). As such, we've just silently corrupted the table descriptor. 

The test fails a bit later, when we're waiting for leases to expire on repartitioned tables. This requires reading the descriptor from the store, which incidentally validates cross references, and therefore expects a partition corresponding to the region dropped in the first operation to be present. 

For me, the key take away here is that we're being overly strict in our pursuit to validate partitions <-> regions on the multi-region enum. I'd say lets loosen this to instead be:
- There must be a partition corresponding to all PUBLIC members. 
- There may or may not be a partition for a READ_ONLY member.
- There should be no partition if there isn't a corresponding enum member (either READ_ONLY or PUBLIC).

Coming back to the aside, I don't think it should be possible for us to get to the point where this test failed. We just corrupted some tables. We should have failed validation when writing the descriptor, not when reading it to refresh leases.  Maybe we need to explicitly validate the locality config here before writing it to the batch?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63459#-:-MY66DONAbL9F78mBfXq:b-up7xqk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)



<!-- Sent from Reviewable.io -->

otan commented:
using #63898 instead
